core/gather/gatherers/root-causes.js

```diff
@@ -0,0 +1,134 @@
+          'DOM.pushNodesByBackendIdsToFrontend', {backendNodeIds});
+        return response.nodeIds;
+      },
+      /** @param {number} nodeId */
+      async getNode(nodeId) {
+        try {
+          const response = await driver.defaultSession.sendCommand('DOM.describeNode', {nodeId});
+          // Why is this always zero? Uh, let's fix it here.
+          response.node.nodeId = nodeId;
+          return response.node;
+        } catch (err) {
+          if (err.message.includes('Could not find node with given id')) {
+            // TODO: when injecting an iframe, the engine gets the node of that frame's document element.
+            // But we don't have a way to access that frame. We just have our default session.
+            // Ex:
+            // node cli http://localhost:10503/shift-attribution.html --quiet --only-audits layout-shifts
+            // To fix we must:
+            // - Change trace engine `getNode` protocol interface to also give frame id
+            // - Expand our driver.targetManager to know how to talk to a session connected to a specific frame
+            // When this is fixed, remove this try/catch.
+            // Note: this could be buggy by giving the wrong node detail if a node id meant for a non-main frame
+            // happens to match one from the main frame ... which is pretty likely...
+            return null;
+          }
+          throw err;
+        }
+      },
+      /** @param {number} nodeId */
+      async getComputedStyleForNode(nodeId) {
+        try {
+          const response = await driver.defaultSession.sendCommand(
+            'CSS.getComputedStyleForNode', {nodeId});
+          return response.computedStyle;
+        } catch {
+          return [];
+        }
+      },
+      /** @param {number} nodeId */
+      async getMatchedStylesForNode(nodeId) {
+        try {
+          const response = await driver.defaultSession.sendCommand(
+            'CSS.getMatchedStylesForNode', {nodeId});
+          return response;
+        } catch {
+          return [];
+        }
+      },
+      /** @param {string} url */
+      // eslint-disable-next-line no-unused-vars
+      async fontFaceForSource(url) {
+        return null;
+      },
+    };
+
+    /** @type {LH.Artifacts.TraceEngineRootCauses} */
+    const rootCauses = {
+      layoutShifts: {},
+    };
+    const rootCausesEngine = new TraceEngine.RootCauses(protocolInterface);
+    const layoutShiftEvents = traceEngineResult.LayoutShifts.clusters.flatMap(c => c.events);
+    for (const event of layoutShiftEvents) {
+      const r = await rootCausesEngine.layoutShifts.rootCausesForEvent(traceEngineResult, event);
+      for (const cause of r.fontChanges) {
+        // TODO: why isn't trace engine unwrapping this promise ...
+        cause.fontFace = await cause.fontFace;
+      }
+      rootCauses.layoutShifts[layoutShiftEvents.indexOf(event)] = r;
+    }
+
+    // await driver.defaultSession.sendCommand('DOM.disable');
```

Comment:
> this all feels very brittle.

Probably because it is when we use the same session for everything.

Looks like we need the CSS domain to be disabled at [this point](https://github.com/GoogleChrome/lighthouse/blob/main/core/gather/gatherers/css-usage.js#L115) so that the protocol will emit events for all existing stylesheets when we call `CSS.enable`.